drivers/stepper/ti/drv8424.c

```diff
@@ -0,0 +1,680 @@
+	if (ret != 0) {
+		LOG_ERR("%s: Failed to reset %s pin (error: %d)", dev->name, pin_name, ret);
+		return ret;
+	}
+
+	/* Set microstep pin */
+	switch (value) {
+	case 0:
+		ret = gpio_pin_set_dt(pin, 0);
+		break;
+	case 1:
+		ret = gpio_pin_set_dt(pin, 1);
+		break;
+	case 2:
+		/* Hi-Z is set by configuring pin as disconnected, not
+		 * all gpio controllers support this.
+		 */
+		ret = gpio_pin_configure_dt(pin, GPIO_DISCONNECTED);
+		break;
+	default:
+		break;
+	}
+
+	if (ret != 0) {
+		LOG_ERR("%s: Failed to set %s pin (error: %d)", dev->name, pin_name, ret);
+		return ret;
+	}
+	return 0;
+}
+
+/*
+ * If microstep setter fails, attempt to recover into previous state.
+ */
+int drv8424_microstep_recovery(const struct device *dev)
+{
+	const struct drv8424_config *config = dev->config;
+	struct drv8424_data *data = dev->data;
+	int ret;
+
+	int m0_value = data->pin_states.m0;
+	int m1_value = data->pin_states.m1;
+
+	ret = drv8424_set_microstep_pin(dev, &config->m0_pin, m0_value, "m0_pin");
+	if (ret != 0) {
+		LOG_ERR("%s: Failed to restore microstep configuration (error: %d)", dev->name,
+			ret);
+		return ret;
+	}
+
+	ret = drv8424_set_microstep_pin(dev, &config->m1_pin, m1_value, "m1_pin");
+	if (ret != 0) {
+		LOG_ERR("%s: Failed to restore microstep configuration (error: %d)", dev->name,
+			ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int drv8424_enable(const struct device *dev, bool enable)
+{
+	int ret;
+	const struct drv8424_config *config = dev->config;
+	struct drv8424_data *data = dev->data;
+	bool has_enable_pin = config->en_pin.port != NULL;
+	bool has_sleep_pin = config->sleep_pin.port != NULL;
+
+	/* Check availability of sleep and enable pins, as these might be hardwired. */
+	if (!has_sleep_pin && !has_enable_pin) {
+		LOG_ERR("%s: Failed to enable/disable device, neither sleep pin nor enable pin are "
+			"available. The device is always on.",
+			dev->name);
+		return -ENOTSUP;
+	}
+
+	if (has_sleep_pin) {
+		ret = gpio_pin_set_dt(&config->sleep_pin, !enable);
+		if (ret != 0) {
+			LOG_ERR("%s: Failed to set sleep_pin (error: %d)", dev->name, ret);
+			return ret;
+		}
+		data->pin_states.sleep = enable ? 0U : 1U;
+	}
+
+	if (has_enable_pin) {
+		ret = gpio_pin_set_dt(&config->en_pin, enable);
+		if (ret != 0) {
+			LOG_ERR("%s: Failed to set en_pin (error: %d)", dev->name, ret);
+			return ret;
+		}
+		data->pin_states.en = enable ? 1U : 0U;
+	}
+
+	data->enabled = enable;
+	if (!enable) {
+		counter_stop(config->counter);
+		data->is_moving = false;
+		gpio_pin_set_dt(&config->step_pin, 0);
+		data->step_signal_high = false;
+	}
+
+	return 0;
+}
+
+static int drv8424_set_counter_frequency(const struct device *dev, uint32_t freq_hz)
+{
+	const struct drv8424_config *config = dev->config;
+	struct drv8424_data *data = dev->data;
+	int ret;
+
+	if (freq_hz == 0) {
+		return -EINVAL;
+	}
+
+	data->counter_top_cfg.ticks = counter_us_to_ticks(config->counter, USEC_PER_SEC / freq_hz);
```

Comment:
consider using `DIV_ROUND_CLOSEST` if you want the closest approximation rather than a floor